File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
  private ListNode Reverse(ListNode head, ListNode tail) {
    ListNode pre = null;
    ListNode current = head;
    ListNode next = head;
    while (current != tail) {
      next = current.next;
      current.next = pre;
      pre = current;
      current = next;
    }
    return pre;
  }
  
  public ListNode ReverseKGroup(ListNode head, int k) {
    ListNode tail = head;
    for (int i = 0; i < k; i++) {
      if (tail == null) return head;
      tail = tail.next;
    }
    var newHead = Reverse(head, tail);
    head.next = ReverseKGroup(tail, k);
    return newHead;
  }
}

File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.go
var tail *ListNode = nil
var fullLength = false
func reverse(head *ListNode, k int) *ListNode {
  if head == nil { return head }
  if head.Next == nil || k == 1 {
    tail = head.Next
    fullLength = (k == 1)
    return head
  }  
  reversed := reverse(head.Next, k - 1)
  head.Next.Next = head
  head.Next = tail
  return reversed
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
  if k == 1 || head == nil {
    return head
  }
  reversed := reverse(head, k)
  if fullLength {
    head.Next = reverseKGroup(tail, k)
  } else {
    reversed = reverse(reversed, k)
  }
  return reversed
}